### ISSUE TITLE
Theme checksum error

### ISSUE THREAD

**Issue by 737simpilot** [CIDRAM/CIDRAM#294]
Upon trying to install a theme I get a checksum error. I tried a fresh install and got the same thing. 

Also, changing the template block page title to "Access Denied" doesn't seem to work.  I still see "Generated by CIDRAM" in the block page template. And I don't see any kind of captcha title that the configuration directive allows you to change. I changed it to other and used "denied" but I don't see anything different on my end. Captchas had a title?!

**Comment by 737simpilot** [comment]
Update:

Not sure if related, but if HCaptcha is used, the block page template is missing CSS or something. Also, it won't play my audio code.



```
<?php $CIDRAM['FieldTemplates']['Output'][] = '<div id="bgaud" style="display:none;"></div> <script language="Javascript"> var aud=document.createElement(\'audio\'),x= "hxxps[://]somesite[.]s3-us-west-2[.]amazonaws[.]com/blockpage[.]mp3" ,z=\'audio/mp3\',a=\'<audio autoplay id="bgsound"><source src="\'+x+\'" type="\'+z+\'"></audio>\',b=\'<object><param name="autostart" value="true"><param name="src" value="\'+x+\'"><param name="autoplay" value="true"><param name="controller" value="false"><embed src="\'+x+\'" controller="false" autoplay="true" autostart="true" type="\'+z+\'" /></object>\',au={audio:aud.play?a:b};document.getElementById(\'bgaud\').innerHTML=au.audio; </script>';
```


I have no idea why that was one line. This markdown crap is pissing me off.  :angry:

**Comment by 737simpilot** [comment]
```
<?php $CIDRAM['FieldTemplates']['Output'][] = '<div id="bgaud" style="display:none;"></div> <script language="Javascript"> var aud=document.createElement(\'audio\'),x= "hxxps[://]somesite[.]s3-us-west-2[.]amazonaws[.]com/blockpage[.]mp3" ,z=\'audio/mp3\',a=\'<audio autoplay id="bgsound"><source src="\'+x+\'" type="\'+z+\'"></audio>\',b=\'<object><param name="autostart" value="true"><param name="src" value="\'+x+\'"><param name="autoplay" value="true"><param name="controller" value="false"><embed src="\'+x+\'" controller="false" autoplay="true" autostart="true" type="\'+z+\'" /></object>\',au={audio:aud.play?a:b};document.getElementById(\'bgaud\').innerHTML=au.audio; </script>';
}
```

**Comment by Maikuolan** [comment]
> Upon trying to install a theme I get a checksum error. I tried a fresh install and got the same thing.

I'll investigate this now. Which particular theme were you trying to install which caused the checksum error? Might be something wrong with the files for that theme since one of my recent commits, or maybe I copied across the wrong checksums for some files. I'll take a look.

> And I don't see any kind of captcha title that the configuration directive allows you to change. I changed it to other and used "denied" but I don't see anything different on my end. Captchas had a title?!

Just the CAPTCHAs which appear for "non-block requests" (basically, if you configure CIDRAM to display CAPTCHAs for "non-blocked requests", in those cases, because it isn't entirely appropriate to say *"Access Denied!"*, and the page is, strictly speaking, not generated from the same template file, I figure that having the option to display a different page title there makes sense; i.e., distinct from cases where we're wanting to tell the end-user that their access has been denied). :-)

**Comment by 737simpilot** [comment]
>Which particular theme were you trying to install which caused the checksum error?

Sorry. It was all of them as far as I know.

**Comment by 737simpilot** [comment]
I must have accidentally closed this issue. AH! :angry:

**Comment by Maikuolan** [comment]
> I must have accidentally closed this issue. AH! 😠

No worries; Happens, and easily fixed. ^^

**Comment by Maikuolan** [comment]
> Sorry. It was all of them as far as I know.

I've taken a quick look, but I didn't notice any obvious problems, or anything which would've caused all the themes to fail updating/installing like that. I'll look a little deeper and do some testing later tonight though.

**Comment by Maikuolan** [comment]
*Possibly* now fixed (I pushed some new changes to the themes tonight, with new checksums included accordingly). Mind giving it a quick try again now and letting me know how it goes? Cheers. :-)

**Comment by jjrueger** [comment]
I too have the same problem and the latest updates don't resolve the issue.  The specific issue seems to be with the captcha html file for each theme.  For example the message for RBI is "theme/rbi – captcha_rbi.html – Checksum error! File rejected!
theme/rbi – Failed to update! +0 bytes | -0 bytes | 0.188".

**Comment by Maikuolan** [comment]
Ah.. Found the problem. My bad. Not sure whether there was some kind of partial reversion or whatever that happened during a recent commit, or whether I simply just outright forgot to copy all the correct files across during a recent commit (because I'm pretty sure I *remember* copying them all across as per needed), but I compared the copies of those files in my local copy of the Git repository for the themes against those in my local dev copy for CIDRAM just now, and they weren't the same (suggesting that either my memory there about copying them across was wrong, or that something of that nature had gone wrong; I don't know). In any case, I've copied across all the correct theme files just now and pushed a new commit to fix the problem. It *should* all be good now (fingers crossed against the unexpected, as always). Give it another whirl now and let me know how it goes. Cheers. :-)

**Comment by jjrueger** [comment]
Yes everything works perfectly now.  All the themes updated without any errors overnight when my cli script ran.  Thanks so much for solving the issue and for the fabulous piece of software!

**Comment by Maikuolan** [comment]
You're most welcome, and I'm glad to hear that everything's working again now. :-)

**Comment by 737simpilot** [comment]
Sorry for the late reply. Issue is partially fixed. In the template_data_block_event_title. This part in the screenshot negates what was offed in the obscured theme that I wanted so that CIDRAM is not displayed at all in the block page. 



![img](https://i.postimg.cc/bwC81MhZ/sfdbgdzrfd.jpg)

**Comment by Maikuolan** [comment]
This'll be resolved when v3 becomes stable due to the new `fields` directive anyhow, but I'll see about modifying `hide_version` for v1+v2 to resolve the problem for now. I'll reply back shortly.

**Comment by 737simpilot** [comment]
Okay. 

And the issue about HCaptcha. The block page with HCaptcha has no CSS or what ever. ReCaptcha is fine however.

**Comment by Maikuolan** [comment]
How bizarre.. I can replicate/confirm the problem at my end. The CSS seems to still be defined properly at the page, but the browser is just completely ignoring it for whatever reason. Trying to figure out what's going on with that now.

**Comment by mikeruss1** [comment]
I use a customised version of template_obscured.html, will this be found OK with the recent changes?

**Comment by Maikuolan** [comment]
> I use a customised version of template_obscured.html, will this be found OK with the recent changes?

Should be.  I'm planning to stop updating the obscured theme at this point anyhow, seeing as alternative arrangements for obscuring version numbers and software identity have been made now (e.g., ability to set custom titles now, current changes to `hide_version` that I'll be pushing tonight, plus the new `fields` stuff for v3 later). That means that the updater won't override the obscured theme with any more changes anymore, but that said, unless I introduce some breaking changes to the templating system (which I don't currently plan to do), it should still work all the same.

**Comment by Maikuolan** [comment]
Found what's causing the problem. >.x

![fff](https://user-images.githubusercontent.com/12571108/158570719-9e511ede-4bdd-4daf-9f85-a5ff2e21e2e9.png)

Investigating possible solutions now.

**Comment by Maikuolan** [comment]
Fixed it. :-)

**Comment by 737simpilot** [comment]
Looks fixed, but just one last issue. Or it may not be an issue. The audio I mentioned above that plays on the block page doesn't work for the HCaptcha template. It does for ReCaptcha though. Looking at the console I see this:


![img](https://i.postimg.cc/YC5R5MKc/fgnfgf.jpg)

May or may not be a big deal. Just interesting that the audio plays with ReCaptcha though. That audio code is that one liner above here in this issue. 

Looks like this was a two to three code line fix. And one line of code was just a whitespace! That's nuts! Probably why I could never code where even a whitespace can take the whole thing down. 

Question: Is there a built in time out for captcha API retrieval? And if so, what does it fallback to if anything? Perhaps if two captchas were configured it'll fall back to one or the other?

Couple things I want to let you know when I updated CIDRAM. 

I used the built-in verification option and got a verify check error for CIDRAM core I think it was. I hit repair and that fixed the issue. 

Verification is failing for BOBUM, however integrity and sanity pass. Not sure what the difference is between verification and integrity. The CRC (?) checks out though. I'm assuming that's the integrity part. I'm also assuming verification fails due to no upstream URL fetch capability?
